File: apps/desktop/src/lib/highlights.ts
// Day-highlight DTOs — the frontend mirror of
// `crates/capture-types/src/highlights.rs`. These are the wire shapes of the
// `get_conversations` / `get_moments` Tauri commands, which are read-time
// queries (no conversations table, no moments table) feeding the Overview tiles
// and the timeline drawer header.
//
// Both commands take a half-open `[startMs, endMs)` window; all timestamps are
// unix milliseconds.

import type { ActivityFocus } from "./types/recording";

/** An Activity whose window overlapped recorded speech for at least two
 *  minutes — the read-time join of activities against speaker turns. */
export interface ConversationCluster {
  activityId: number;
  title: string;
  startedAtMs: number;
  endedAtMs: number;
  /** Total speech overlap inside the Activity window, summed across turns.
   *  Always >= the backend's two-minute threshold. */
  spokenMs: number;
  /** Distinct speaker clusters heard inside the window; 1 is valid. */
  speakerCount: number;
}

/** A headline frame of an Activity, ranked by focus band then Activity
 *  duration. `filePath` is an on-disk path — render it via `convertFileSrc`. */
export interface Moment {
  frameId: number;
  filePath: string;
  capturedAtMs: number;
  activityId: number;
  title: string;
  /** Effective focus band (a user correction wins); null when unlabelled. */
  focus: ActivityFocus | null;
  activityStartedAtMs: number;
  activityEndedAtMs: number;
  /** `activityEndedAtMs - activityStartedAtMs`. */
  durationMs: number;
}

File: apps/desktop/src-tauri/src/highlights.rs
//! Tauri command surface for the read-time **day highlights**: conversations
//! and moments.
//!
//! Thin adapter over `app_infra::HighlightsStore` (CLAUDE.md boundary: app-infra
//! owns SQLite, Tauri handlers stay thin). No aggregation here — it forwards the
//! half-open `[startMs, endMs)` window and maps the store error onto the
//! `Result<_, String>` Tauri seam.

use capture_types::{ConversationCluster, Moment};

use crate::app_infra::AppInfraState;

/// Activities in `[start_ms, end_ms)` whose window overlaps recorded speech for
/// at least two minutes, newest first, with the number of distinct speakers
/// heard in each.
#[tauri::command]
pub async fn get_conversations(
    infra: tauri::State<'_, AppInfraState>,
    start_ms: i64,
    end_ms: i64,
) -> Result<Vec<ConversationCluster>, String> {
    infra
        .highlights()
        .conversations(start_ms, end_ms)
        .await
        .map_err(|e| e.to_string())
}

/// Headline frames of Activities in `[start_ms, end_ms)`, ranked by the
/// Activity's focus band then its duration. `limit` defaults to the store's
/// default when omitted.
#[tauri::command]
pub async fn get_moments(
    infra: tauri::State<'_, AppInfraState>,
    start_ms: i64,
    end_ms: i64,
    limit: Option<i64>,
) -> Result<Vec<Moment>, String> {
    infra
        .highlights()
        .moments(start_ms, end_ms, limit)
        .await
        .map_err(|e| e.to_string())
}
